spark/src/main/scala/org/apache/spark/sql/delta/TypeWideningMetadata.scala

```diff
@@ -0,0 +1,192 @@
+    }
+    builder.build()
+  }
+}
+
+private[delta] object TypeChange {
+  val TABLE_VERSION_METADATA_KEY: String = "tableVersion"
+  val FROM_TYPE_METADATA_KEY: String = "fromType"
+  val TO_TYPE_METADATA_KEY: String = "toType"
+  val FIELD_PATH_METADATA_KEY: String = "fieldPath"
+
+   /** Deserialize this type change from a [[Metadata]] object. */
+  def fromMetadata(metadata: Metadata): TypeChange = {
+    val fieldPath = if (metadata.contains(FIELD_PATH_METADATA_KEY)) {
+      metadata.getString(FIELD_PATH_METADATA_KEY).split("\\.").toSeq
+    } else {
+      Seq.empty
+    }
+    TypeChange(
+      version = metadata.getLong(TABLE_VERSION_METADATA_KEY),
+      fromType = DataType.fromDDL(metadata.getString(FROM_TYPE_METADATA_KEY)),
+      toType = DataType.fromDDL(metadata.getString(TO_TYPE_METADATA_KEY)),
+      fieldPath
+    )
+  }
+}
+
+/**
+ * Represents all type change information for a single struct field
+ * @param typeChanges The type changes that have been applied to the field.
+ */
+private[delta] case class TypeWideningMetadata(typeChanges: Seq[TypeChange]) {
+
+  import TypeWideningMetadata._
+
+  /**
+   * Add the type changes to the metadata of the given field, preserving any pre-existing type
+   * widening metadata.
+   */
+  def appendToField(field: StructField): StructField = {
+    if (typeChanges.isEmpty) return field
+
+    val existingTypeChanges = fromField(field).map(_.typeChanges).getOrElse(Seq.empty)
+    val allTypeChanges = existingTypeChanges ++ typeChanges
+
+    val newMetadata = new MetadataBuilder().withMetadata(field.metadata)
+      .putMetadataArray(TYPE_CHANGES_METADATA_KEY, allTypeChanges.map(_.toMetadata).toArray)
+      .build()
+    field.copy(metadata = newMetadata)
+  }
+}
+
+private[delta] object TypeWideningMetadata {
+  val TYPE_CHANGES_METADATA_KEY: String = "delta.typeChanges"
+
+  /** Read the type widening metadata from the given field. */
+  def fromField(field: StructField): Option[TypeWideningMetadata] = {
+    Option.when(field.metadata.contains(TYPE_CHANGES_METADATA_KEY)) {
+      val typeChanges = field.metadata.getMetadataArray(TYPE_CHANGES_METADATA_KEY)
+        .map { changeMetadata =>
+          TypeChange.fromMetadata(changeMetadata)
+        }.toSeq
+      TypeWideningMetadata(typeChanges)
+    }
+  }
+
+  /**
+   * Computes the type changes from `oldSchema` to `schema` and adds corresponding type change
+   * metadata to `schema`.
+   */
+  def addTypeWideningMetadata(
+      txn: OptimisticTransaction,
+      schema: StructType,
+      oldSchema: StructType): StructType = {
+
+    if (!TypeWidening.isEnabled(txn.protocol, txn.metadata)) return schema
+
+    if (DataType.equalsIgnoreNullability(schema, oldSchema)) return schema
+
+    SchemaMergingUtils.transformColumns(schema, oldSchema) {
+      case (_, newField, Some(oldField), _) =>
+        val typeChanges =
+          collectTypeChanges(oldField.dataType, newField.dataType, txn.getFirstAttemptVersion)
```

Comment:
Is the getFirstAttemptVersion ever updated in the case of concurrent transactions on the table? Maybe add a comment?

Reply:
I added a comment